crates/astria-core/src/primitive/v1/asset/denom.rs

```diff
@@ -0,0 +1,554 @@
+                return Err(Self::Err::channel_is_empty());
+            }
+            parsed_segments.push(PortAndChannel {
+                port: port.into(),
+                channel: channel.into(),
+            });
+        }
+        Ok(parsed_segments)
+    }
+}
+
+impl TraceSegments {
+    fn new() -> Self {
+        Self {
+            inner: VecDeque::new(),
+        }
+    }
+
+    fn push(&mut self, seg: PortAndChannel) {
+        self.inner.push_back(seg);
+    }
+
+    fn pop_first_port_and_channel(&mut self) -> Option<PortAndChannel> {
+        self.inner.pop_front()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct PortAndChannel {
+    port: String,
+    channel: String,
+}
+
+impl PortAndChannel {
+    fn channel(&self) -> &str {
+        &self.channel
+    }
+
+    fn port(&self) -> &str {
+        &self.port
+    }
+}
+
+impl std::fmt::Display for TracePrefixed {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for segment in &self.trace.inner {
+            f.write_str(&segment.port)?;
+            f.write_str("/")?;
+            f.write_str(&segment.channel)?;
+            f.write_str("/")?;
+        }
+        f.write_str(&self.base_denom)
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error(transparent)]
+pub struct ParseTracePrefixedError(ParseTracePrefixedErrorKind);
+
+impl ParseTracePrefixedError {
+    fn base_is_empty() -> Self {
+        Self(ParseTracePrefixedErrorKind::BaseIsEmpty)
+    }
+
+    fn channel_is_empty() -> Self {
+        Self(ParseTracePrefixedErrorKind::ChannelIsEmpty)
+    }
+
+    fn port_is_empty() -> Self {
+        Self(ParseTracePrefixedErrorKind::PortIsEmpty)
+    }
+
+    fn not_ascii() -> Self {
+        Self(ParseTracePrefixedErrorKind::NotAscii)
+    }
+
+    fn port_without_channel() -> Self {
+        Self(ParseTracePrefixedErrorKind::PortWithoutChannel)
+    }
+
+    fn whitespace() -> Self {
+        Self(ParseTracePrefixedErrorKind::Whitespace)
+    }
+}
+
+#[derive(Debug, thiserror::Error, PartialEq)]
+enum ParseTracePrefixedErrorKind {
+    #[error("the input itself or its base denom segment is empty")]
+    BaseIsEmpty,
+    #[error("a port segment was empty")]
+    PortIsEmpty,
+    #[error("a channel segment was empty")]
+    ChannelIsEmpty,
+    #[error("input contained non-ascii characters")]
+    NotAscii,
+    #[error("input contains whitespace")]
+    Whitespace,
+    #[error(
+        "the denom trace path was lopsided, there was one port without matching channel segment"
+    )]
+    PortWithoutChannel,
+}
+
+impl FromStr for TracePrefixed {
+    type Err = ParseTracePrefixedError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if !s.is_ascii() {
+            return Err(Self::Err::not_ascii());
+        }
+        if s.as_bytes().iter().any(u8::is_ascii_whitespace) {
+            return Err(Self::Err::whitespace());
+        }
+        let (base_denom, segments) = match s.rsplit_once('/') {
```

Comment:
i think this is reversed, should be `let (segments, base_denom) = match s.rsplit_once('/')` since the base denom is always after the last slash

Reply:
can you add a unit test for this?

Reply:
Admittedly, the order is confusing and I will change it, but the important bit is the body of the match statement.

This is indirectly unit tested through the parse-logic below. But I will add a more explicit unit test that actually checks the fields being set correctly.

Reply:
Switched the var. Looked at the unit tests: the `pop_path` test covers all unit test needs for the parse logic I think.

 010b7e